### PR TITLE
[13.0] [MIG] web_widget_numeric_step: Migration to v13

### PR DESCRIPTION
**Pull request by CarlosRoca13**
cc @Tecnativa TT27480

please @Tardo @sergio-teruel review this.

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+  "restructuredtext.confPath": ""
+}
```
Comment (Tardo) [inline .vscode/settings.json]: :fire: this file

**Review by Tardo — CHANGES_REQUESTED** [review]

**Comment by CarlosRoca13** [comment]
@Tardo changes done

File: web_widget_numeric_step/static/src/js/numeric_step.js [hunk 2]
```diff
@@ -0,0 +1,329 @@
+/* Copyright 2019 GRAP - Quentin DUPONT
+ * Copyright 2020 Tecnativa - Alexandre Díaz
+ * License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html) */
+
+odoo.define("web_widget_numeric_step.field", function(require) {
+    "use strict";
+
+    var field_utils = require("web.field_utils");
+    var Registry = require("web.field_registry");
+    var FieldFloat = require("web.basic_fields").FieldFloat;
```
Comment (Tardo) [inline web_widget_numeric_step/static/src/js/numeric_step.js]:
```suggestion
    const field_utils = require("web.field_utils");
    const Registry = require("web.field_registry");
    const FieldFloat = require("web.basic_fields").FieldFloat;
```

File: web_widget_numeric_step/static/src/js/numeric_step.js [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+
+    var NumericStep = FieldFloat.extend({
```
Comment (Tardo) [inline web_widget_numeric_step/static/src/js/numeric_step.js]:
```suggestion
    const NumericStep = FieldFloat.extend({
```

File: web_widget_numeric_step/static/src/js/numeric_step.js [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+
+        /**
+         * @override
+         */
+        init: function() {
+            this._super.apply(this, arguments);
+
+            // Widget config
+            var max_val = this.nodeOptions.max;
+            var min_val = this.nodeOptions.min;
```
Comment (Tardo) [inline web_widget_numeric_step/static/src/js/numeric_step.js]:
```suggestion
            let max_val = this.nodeOptions.max;
            let min_val = this.nodeOptions.min;
```

File: web_widget_numeric_step/static/src/js/numeric_step.js [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+            this._config = {
+                step: Number(this.nodeOptions.step) || 1,
+                min: Number(min_val),
+                max: Number(max_val),
+            };
+
+            var _this = this;
+            this._lazyOnChangeTrigger = _.debounce(function() {
+                _this.$input.trigger("change");
+            }, this.DELAY_THROTTLE_CHANGE);
```
Comment (Tardo) [inline web_widget_numeric_step/static/src/js/numeric_step.js]:
```suggestion
            this._lazyOnChangeTrigger = _.debounce(
                () => this.$input.trigger("change"),
                this.DELAY_THROTTLE_CHANGE);
```

File: web_widget_numeric_step/static/src/js/numeric_step.js [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+                    "mouseup",
+                    $.proxy(_this, "_onMouseUp"),
+                    false
+                );
+                document.addEventListener(
+                    "touchend",
+                    $.proxy(_this, "_onMouseUp"),
+                    false
+                );
+            });
```
Comment (Tardo) [inline web_widget_numeric_step/static/src/js/numeric_step.js]:
```suggestion
            this._click_delay = this.DEF_CLICK_DELAY;
            this._autoStep = false;
            return this._super.apply(this, arguments).then(() => {
                document.addEventListener(
                    "mouseup",
                    this._onMouseUp,
                    false
                );
                document.addEventListener(
                    "touchend",
                    this._onMouseUp,
                    false
                );
            });
```

File: web_widget_numeric_step/static/src/js/numeric_step.js [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+            return this._super.apply(this, arguments);
+        },
+
+        /**
+         * Transform widget value to usable database value
+         *
+         * @override
+         */
+        _parseValue: function() {
+            var parsedVal = this._super.apply(this, arguments);
```
Comment (Tardo) [inline web_widget_numeric_step/static/src/js/numeric_step.js]:
```suggestion
            const parsedVal = this._super.apply(this, arguments);
```

File: web_widget_numeric_step/static/src/js/numeric_step.js [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+            return parsedVal;
+        },
+
+        /**
+         * Adds HTML attributes to the input
+         *
+         * @override
+         */
+        _prepareInput: function() {
+            var result = this._super.apply(this, arguments);
```
Comment (Tardo) [inline web_widget_numeric_step/static/src/js/numeric_step.js]:
```suggestion
            const result = this._super.apply(this, arguments);
```

File: web_widget_numeric_step/static/src/js/numeric_step.js [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+         * @override
+         */
+        _renderEdit: function() {
+            _.defer(
+                function() {
+                    this.$el
+                        .parents("td.o_numeric_step_cell")
+                        .addClass("numeric_step_editing_cell");
+                }.bind(this)
+            );
```
Comment (Tardo) [inline web_widget_numeric_step/static/src/js/numeric_step.js]:
```suggestion
            _.defer(
                () => this.$el
                        .parents("td.o_numeric_step_cell")
                        .addClass("numeric_step_editing_cell")
            );
```

File: web_widget_numeric_step/static/src/js/numeric_step.js [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+            this._super.apply(this, arguments);
+        },
+
+        /**
+         * Increase/Decrease widget input value
+         *
+         * @param {String} mode can be "plus" or "minus"
+         */
+        _doStep: function(mode) {
+            var cval = 0;
```
Comment (Tardo) [inline web_widget_numeric_step/static/src/js/numeric_step.js]:
```suggestion
            let cval = 0;
```

File: web_widget_numeric_step/static/src/js/numeric_step.js [hunk 11]
```diff
@@ -0,0 +1,329 @@
...
+            try {
+                var field = this.record.fields[this.name];
```
Comment (Tardo) [inline web_widget_numeric_step/static/src/js/numeric_step.js]:
```suggestion
                const field = this.record.fields[this.name];
```

File: web_widget_numeric_step/static/src/js/numeric_step.js [hunk 12]
```diff
@@ -0,0 +1,329 @@
...
+            } catch (e) {
+                cval = this.value;
+                mode = false; // Only set the value in this case
+            }
+            if (mode === "plus") {
+                cval += this._config.step;
+            } else if (mode === "minus") {
+                cval -= this._config.step;
+            }
+            var nval = this._sanitizeNumberValue(cval);
```
Comment (Tardo) [inline web_widget_numeric_step/static/src/js/numeric_step.js]:
```suggestion
            const nval = this._sanitizeNumberValue(cval);
```

File: web_widget_numeric_step/static/src/js/numeric_step.js [hunk 13]
```diff
@@ -0,0 +1,329 @@
...
+
+        // Handle Events
+
+        /**
+         * @private
+         * @param {MouseClickEvent} ev
+         */
+        _onStepClick: function(ev) {
+            if (!this._autoStep) {
+                var mode = ev.target.dataset.mode;
```
Comment (Tardo) [inline web_widget_numeric_step/static/src/js/numeric_step.js]:
```suggestion
                const mode = ev.target.dataset.mode;
```

File: web_widget_numeric_step/static/src/js/numeric_step.js [hunk 14]
```diff
@@ -0,0 +1,329 @@
...
+        /**
+         * @private
+         * @param {MouseClickEvent} ev
+         */
+        _onStepMouseDown: function(ev) {
+            if (ev.button === 0 && !this._auto_step_interval) {
+                this._auto_step_interval = setTimeout(
+                    $.proxy(this, "_whileMouseDown", ev),
+                    this._click_delay
+                );
```
Comment (Tardo) [inline web_widget_numeric_step/static/src/js/numeric_step.js]:
```suggestion
                this._auto_step_interval = setTimeout(
                    this._whileMouseDown.bind(this, ev),
                    this._click_delay
                );
```

File: web_widget_numeric_step/static/src/js/numeric_step.js [hunk 15]
```diff
@@ -0,0 +1,329 @@
...
+            this._clearStepInterval();
+        },
+
+        /**
+         * @private
+         * @param {MouseClickEvent} ev
+         */
+        _whileMouseDown: function(ev) {
+            this._autoStep = true;
+            var mode = ev.target.dataset.mode;
```
Comment (Tardo) [inline web_widget_numeric_step/static/src/js/numeric_step.js]:
```suggestion
            const mode = ev.target.dataset.mode;
```

File: web_widget_numeric_step/static/src/js/numeric_step.js [hunk 16]
```diff
@@ -0,0 +1,329 @@
...
+        /**
+         * Check limits and precision of the value.
+         * If the value 'is not a number', the function does nothing to
+         * be good with other possible modules.
+         *
+         * @param {Number} value
+         * @returns {Number}
+         */
+        _sanitizeNumberValue: function(value) {
+            var cval = Number(value);
```
Comment (Tardo) [inline web_widget_numeric_step/static/src/js/numeric_step.js]:
```suggestion
            let cval = Number(value);
```

File: web_widget_numeric_step/static/src/js/numeric_step.js [hunk 17]
```diff
@@ -0,0 +1,329 @@
...
+            if (_.isNaN(cval)) {
+                return value;
+            }
+            if (!_.isNaN(this._config.min) && cval < this._config.min) {
+                cval = this._config.min;
+            } else if (!_.isNaN(this._config.max) && cval > this._config.max) {
+                cval = this._config.max;
+            }
+
+            var field = this.record.fields[this.name];
```
Comment (Tardo) [inline web_widget_numeric_step/static/src/js/numeric_step.js]:
```suggestion
            const field = this.record.fields[this.name];
```

File: web_widget_numeric_step/static/src/js/numeric_step.js [hunk 18]
```diff
@@ -0,0 +1,329 @@
...
+            var formattedValue = field_utils.format[field.type](cval, field, {
+                data: this.record.data,
+                escape: true,
+                isPassword: false,
+                digits: field.digits,
+            });
+            return formattedValue;
```
Comment (Tardo) [inline web_widget_numeric_step/static/src/js/numeric_step.js]:
```suggestion
            // formatted value
            return field_utils.format[field.type](cval, field, {
                data: this.record.data,
                escape: true,
                isPassword: false,
                digits: field.digits,
            });
```

**Review by Tardo — CHANGES_REQUESTED** [review]
Try update js to use ES6 :)

**Comment by CarlosRoca13** [comment]
@Tardo changes done

File: web_widget_numeric_step/static/src/js/numeric_step.js [hunk 19]
```diff
@@ -0,0 +1,318 @@
...
+         * Add global events listeners
+         *
+         * @override
+         */
+        start: function() {
+            this._click_delay = this.DEF_CLICK_DELAY;
+            this._autoStep = false;
+            return this._super.apply(this, arguments).then(() => {
+                document.addEventListener("mouseup", this._onMouseUp, false);
+                document.addEventListener("touchend", this._onMouseUp, false);
```
Comment (Tardo) [inline web_widget_numeric_step/static/src/js/numeric_step.js]:
```suggestion
                document.addEventListener("mouseup", this._onMouseUp.bind(this), false);
                document.addEventListener("touchend", this._onMouseUp.bind(this), false);
```

**Review by Tardo — CHANGES_REQUESTED** [review]

**Comment by CarlosRoca13** [comment]
@Tardo changed

**Review by Tardo — APPROVED** [review]
LGTM :+1: Thanks @CarlosRoca13

**Review by pedrobaeza — APPROVED** [review]

**Comment by lmignon** [comment]
@pedrobaeza AFAIK it's important to bump the version when merging PR of migrated addon. Without bump the bot will not publish a new release of the addon, only a dev release...

**Comment by pedrobaeza** [comment]
I don't like to have the initial version of a module as `13.0.1.0.1`. That requirement is very artificial, and due to all the overhead requires for the pip thing, that I don't share. If that's a problem for that part, I think the fix should come from the publishing method.

**Pull request merged**